instrumentation/mongo/mongo-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/mongo/v4_0/MongoClientInstrumentationModule.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.mongo.v4_0;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonMap;
+import static net.bytebuddy.matcher.ElementMatchers.declaresMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.event.CommandListener;
+import com.mongodb.internal.async.SingleResultCallback;
+import io.opentelemetry.javaagent.instrumentation.api.Java8BytecodeBridge;
+import io.opentelemetry.javaagent.instrumentation.mongo.TracingCommandListener;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.lang.reflect.Modifier;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(InstrumentationModule.class)
+public class MongoClientInstrumentationModule extends InstrumentationModule {
+
+  public MongoClientInstrumentationModule() {
+    super("mongo", "mongo-4.0");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return asList(
+        new MongoClientSettingsBuilderInstrumentation(),
+        new InternalStreamConnectionInstrumentation(),
+        new BaseClusterInstrumentation());
+  }
+
+  private static final class MongoClientSettingsBuilderInstrumentation
+      implements TypeInstrumentation {
+    @Override
+    public ElementMatcher<TypeDescription> typeMatcher() {
+      return named("com.mongodb.MongoClientSettings$Builder")
+          .and(
+              declaresMethod(
+                  named("addCommandListener")
+                      .and(
+                          takesArguments(
+                              new TypeDescription.Latent(
+                                  "com.mongodb.event.CommandListener",
+                                  Modifier.PUBLIC,
+                                  null,
+                                  Collections.<TypeDescription.Generic>emptyList())))
```

Comment:
Why not just `named("com.mongodb.event.CommandListener")`?

Reply:
This is copy paste from mongo 3.7, I have no idea why it is done this way. I'll change it to use `named` for all mongo versions.